### PR TITLE
Ia 59 fix floataing stat for enrollment

### PR DESCRIPTION
**Pull request by flying-pi**
Provide correct statistic for the enrollment tab. 

Add celerybeat for the update cache of the enrollment statistic.

### CODE REVIEW

File: rg_instructor_analytics/models.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+class EnrollmentByStudent(Model):
+    """
+    Model for store last enrollment state of a user.
+    """
+
+    course_id = CourseKeyField(max_length=255, db_index=True)
+    student = ForeignKey(User, db_index=True)
+    last_update = DateTimeField(auto_now_add=False, db_index=True)
+    state = PositiveSmallIntegerField()
```
Comment (idegtiarov) [inline rg_instructor_analytics/models.py]: You can use boolean field here

**Review by idegtiarov — CHANGES_REQUESTED** [review]

File: rg_instructor_analytics/models.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+    class Meta:
+        """
+        Meta class.
+        """
+
+        unique_together = ('course_id', 'student',)
+
+
+class TotalEnrollmentByCourse(Model):
```
Comment (idegtiarov) [inline rg_instructor_analytics/models.py]: It seems you can avoid this model, and use the first one instead.

File: rg_instructor_analytics/tasks.py [hunk 3]
```diff
@@ -23,3 +36,90 @@ def send_email_to_cohort(subject, message, students):
     msg.encoding = 'UTF-8'
     msg.attach_alternative(html_content, "text/html")
     msg.send(fail_silently=False)
+
+
+cron_settings = getattr(
+    settings, 'RG_ANALYTICS_ENROLLMENT_STAT_UPDATE',
+    {
+        'minute': '*/1',
+        'hour': '*',
```
Comment (idegtiarov) [inline rg_instructor_analytics/tasks.py]: I would leave the only `minute` key as default value

File: rg_instructor_analytics/tasks.py [hunk 4]
```diff
@@ -23,3 +36,90 @@ def send_email_to_cohort(subject, message, students):
...
+        'day_of_month': '*',
+        'month_of_year': '*',
+    }
+)
+
+
+@periodic_task(run_every=crontab(**cron_settings))
+def enrollment_collector_date():
+    """
+    Task for update enrollment cache.
```
Comment (idegtiarov) [inline rg_instructor_analytics/tasks.py]: Please refrase

File: rg_instructor_analytics/tasks.py [hunk 5]
```diff
@@ -23,3 +36,90 @@ def send_email_to_cohort(subject, message, students):
...
+    """
+    last_cachce = EnrollmentByStudent.objects.all().order_by('-last_update')
```
Comment (idegtiarov) [inline rg_instructor_analytics/tasks.py]: `cachce`? what does it mean?

File: rg_instructor_analytics/tasks.py [hunk 6]
```diff
@@ -23,3 +36,90 @@ def send_email_to_cohort(subject, message, students):
...
+    if last_cachce.exists():
+        last_update = last_cachce.first().last_update
+    else:
+        last_update = datetime(2000, 01, 01, 0, 0)
```
Comment (idegtiarov) [inline rg_instructor_analytics/tasks.py]: Please use general int format

File: rg_instructor_analytics/tasks.py [hunk 7]
```diff
@@ -23,3 +36,90 @@ def send_email_to_cohort(subject, message, students):
...
+    enrollments_history = (
+        CourseEnrollment.history
+        .filter(~Q(history_type='+'))
+        .filter(history_date__gt=last_update)
+        .values("history_date", "is_active", "user", "course_id")
+        .order_by('history_date')
+    )
+    users_state = {
+        (u['student'], u['course_id']): {'last_update': u['last_update'], 'state': u['state']}
+        for u in EnrollmentByStudent.objects.all().values("course_id", "student", "last_update", "state", )
```
Comment (idegtiarov) [inline rg_instructor_analytics/tasks.py]: why `u`?

File: rg_instructor_analytics/tasks.py [hunk 8]
```diff
@@ -23,3 +36,90 @@ def send_email_to_cohort(subject, message, students):
...
+    }
+    total_stat = {
+        t['course_id']: t['total'] for t in TotalEnrollmentByCourse.objects.all().values("course_id", "total", )
+    }
+    result_stat = {}
+    for e in enrollments_history:
```
Comment (idegtiarov) [inline rg_instructor_analytics/tasks.py]: I would avoid one char variables

File: rg_instructor_analytics/tasks.py [hunk 9]
```diff
@@ -23,3 +36,90 @@ def send_email_to_cohort(subject, message, students):
...
+        key = e['user'], e['course_id']
+        if key in users_state and users_state[key]['state'] == e['is_active']:
+            continue
+        users_state[key] = {
+            'last_update': e['history_date'],
+            'state': e['is_active'],
+        }
+        total_key = e['history_date'].date(), e['course_id']
```
Comment (idegtiarov) [inline rg_instructor_analytics/tasks.py]: I would use parentheses with the tuple definition

**Review by idegtiarov — CHANGES_REQUESTED** [review]

File: rg_instructor_analytics/models.py (hunk 1 above)
Comment (flying-pi) [reply rg_instructor_analytics/models.py]: done

File: rg_instructor_analytics/models.py (hunk 2 above)
Comment (flying-pi) [reply rg_instructor_analytics/models.py]: done

File: rg_instructor_analytics/tasks.py (hunk 3 above)
Comment (flying-pi) [reply rg_instructor_analytics/tasks.py]: done

File: rg_instructor_analytics/tasks.py (hunk 4 above)
Comment (flying-pi) [reply rg_instructor_analytics/tasks.py]: done

File: rg_instructor_analytics/tasks.py (hunk 5 above)
Comment (flying-pi) [reply rg_instructor_analytics/tasks.py]: fix.

File: rg_instructor_analytics/tasks.py (hunk 6 above)
Comment (flying-pi) [reply rg_instructor_analytics/tasks.py]: done

File: rg_instructor_analytics/tasks.py (hunk 7 above)
Comment (flying-pi) [reply rg_instructor_analytics/tasks.py]: fix

File: rg_instructor_analytics/tasks.py (hunk 8 above)
Comment (flying-pi) [reply rg_instructor_analytics/tasks.py]: ok, I fixed that.

File: rg_instructor_analytics/tasks.py (hunk 9 above)
Comment (flying-pi) [reply rg_instructor_analytics/tasks.py]: fix

File: rg_instructor_analytics/models.py [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+
+
+class EnrollmentByStudent(Model):
+    """
+    Model for store last enrollment state of a user.
+    """
+
+    course_id = CourseKeyField(max_length=255, db_index=True)
+    student = ForeignKey(User, db_index=True)
+    last_update = DateTimeField(auto_now_add=False, db_index=True)
```
Comment (idegtiarov) [inline rg_instructor_analytics/models.py]: I would avoid using the default parameters

File: rg_instructor_analytics/models.py (hunk 2 above)
Comment (idegtiarov) [reply rg_instructor_analytics/models.py]: Or if you prefer to persist this data please use the cache.

**Review by idegtiarov — CHANGES_REQUESTED** [review]

File: rg_instructor_analytics/models.py (hunk 2 above)
Comment (flying-pi) [reply rg_instructor_analytics/models.py]: done.

File: rg_instructor_analytics/models.py (hunk 10 above)
Comment (flying-pi) [reply rg_instructor_analytics/models.py]: done

**Review by idegtiarov — APPROVED** [review]

**Pull request merged**